test/jasmine/tests/plot_api_test.js

```diff
@@ -3211,86 +3217,127 @@ describe('Test plot api', function() {
                 }
 
-                // Make sure we define `_length` in every trace *in supplyDefaults*.
-                // This is only relevant for traces that *have* a 1D concept of length,
-                // and in addition to simplifying calc/plot logic later on, ths serves
-                // as a signal to transforms about how they should operate. For traces
-                // that do NOT have a 1D length, `_length` should be `null`.
-                var mockGD = Lib.extendDeep({}, mock);
-                supplyAllDefaults(mockGD);
-                expect(mockGD._fullData.length).not.toBeLessThan((mock.data || []).length, mockSpec[0]);
-                mockGD._fullData.forEach(function(trace, i) {
-                    var len = trace._length;
-                    if(trace.visible !== false && len !== null) {
-                        expect(typeof len).toBe('number', mockSpec[0] + ' trace ' + i + ': type=' + trace.type);
-                    }
-                });
+                typesTested[trace.type]++;
 
-                Plotly.newPlot(gd, mock)
-                .then(countPlots)
-                .then(function() {
-                    initialJson = fullJson();
-
-                    return Plotly.react(gd, mock);
-                })
-                .then(function() {
-                    expect(fullJson()).toEqual(initialJson);
-                    countCalls({});
-                })
-                .catch(failTest)
-                .then(done);
+                if(trace.transforms) {
+                    trace.transforms.forEach(function(transform) {
+                        typesTested[transform.type]++;
+                    });
+                }
             });
+
+            Plotly.newPlot(gd, mock)
+            .then(countPlots)
+            .then(function() {
+                initialJson = fullJson();
+
+                return Plotly.react(gd, mock);
+            })
+            .then(function() {
+                expect(fullJson()).toEqual(initialJson);
+                countCalls({});
+            })
+            .catch(failTest)
+            .then(done);
+        }
+
+        mockList.forEach(function(mockSpec) {
+            it('can redraw "' + mockSpec[0] + '" with no changes as a noop', function(done) {
+                _runReactMock(mockSpec, done);
+            });
+        });
+
+        it('@noCI can redraw scattermapbox with no changes as a noop', function(done) {
+            typesTested.scattermapbox = 0;
+
+            Plotly.setPlotConfig({
+                mapboxAccessToken: require('@build/credentials.json').MAPBOX_ACCESS_TOKEN
+            });
+
+            _runReactMock(['scattermapbox', require('@mocks/mapbox_bubbles-text.json')], done);
+        });
+
+        it('tested every trace & transform type at least once', function() {
+            for(var itemType in typesTested) {
+                expect(typesTested[itemType]).toBeGreaterThan(0, itemType + ' was not tested');
+            }
```

Comment:
So when a new trace or transform type is added we'll be forced to add it to this suite as well. (note above I excluded `contourgl` and, on CI, `scattermapbox`)

I couldn't think of a good way to ensure that every *component* gets tested as well... in fact looking over the list now I don't see shapes or 3D annotations, I'll have to add those. If anyone has ideas how to enforce this please chime in!

Reply:
OK - still sounds a bit finicky... perhaps a later iteration we could add a test that ensures we've covered every attribute anywhere in the schema - that would be a significantly more stringent coverage test for traces and transforms as well! But if you don't mind I'll defer this for now.

Reply:
> But if you don't mind I'll defer this for now.

No problem :+1: